### PR TITLE
Feature toggleability

### PR DESCRIPTION
**Pull request by msohailhussain**

### CODE REVIEW

File: core-api/src/test/java/com/optimizely/ab/OptimizelyTest.java [hunk 1]
```diff
@@ -3472,6 +3472,79 @@ public void isFeatureEnabledReturnsTrueButDoesNotSendWhenUserIsBucketedIntoVaria
...
+                .withConfig(validProjectConfig)
+                .withErrorHandler(mockErrorHandler)
+                .build();
+
+        optimizely.setForcedVariation(activatedExperiment.getKey(), testUserId, forcedVariation.getKey() );
+        assertFalse(optimizely.isFeatureEnabled(FEATURE_FLAG_MULTI_VARIATE_FEATURE.getKey(), testUserId));
+    }
+
+    @Test
+    public void isFeatureEnabledTrueWhenFeatureEnabledOfVariationIsTrue() throws Exception{
```
Comment (wangjoshuah) [inline core-api/src/test/java/com/optimizely/ab/OptimizelyTest.java]: describe the test case in the comments.

Comment (wangjoshuah) [inline core-api/src/test/java/com/optimizely/ab/OptimizelyTest.java]: describe the test case in in comments above

File: core-api/src/test/java/com/optimizely/ab/OptimizelyTest.java [hunk 2]
```diff
@@ -3472,6 +3472,79 @@ public void isFeatureEnabledReturnsTrueButDoesNotSendWhenUserIsBucketedIntoVaria
...
+                eq(genericUserId),
+                eq(Collections.<String, String>emptyMap())
+        );
+
+        assertTrue(spyOptimizely.isFeatureEnabled(validFeatureKey, genericUserId));
+
+    }
+
+    @Test
+    public void isFeatureEnabledFalseWhenFeatureEnabledOfVariationIsFalse() throws Exception{
```
Comment (wangjoshuah) [inline core-api/src/test/java/com/optimizely/ab/OptimizelyTest.java]: add test description here.

**Comment by wangjoshuah** [comment]
build

**Comment by mikeng13** [comment]
build

File: core-api/src/main/java/com/optimizely/ab/config/Variation.java [hunk 3]
```diff
@@ -37,19 +37,28 @@
...
+    }
+
     @JsonCreator
     public Variation(@JsonProperty("id") String id,
                      @JsonProperty("key") String key,
+                     @JsonProperty("featureEnabled") Boolean featureEnabled,
                      @JsonProperty("variables") List<LiveVariableUsageInstance> liveVariableUsageInstances) {
         this.id = id;
         this.key = key;
+        this.featureEnabled = featureEnabled;
```
Comment (thomaszurkan-optimizely) [inline core-api/src/main/java/com/optimizely/ab/config/Variation.java]: I see in the JSON parsers you make sure that the featureEnabled is not null.  Can't you do that here as well?  if (featureEnbaled == null)  set it to false.  Just for safety sake.

File: core-api/src/test/java/com/optimizely/ab/OptimizelyTest.java (hunk 2 above)
Comment (thomaszurkan-optimizely) [reply core-api/src/test/java/com/optimizely/ab/OptimizelyTest.java]: Can we add one more explicit test: isFeatureEnabledFalseWhenFeatureEnabledOfVariationIsNotPresentInJson ?
Feel free to not use that name.

**Comment by mikeng13** [comment]
build

**Comment by wangjoshuah** [comment]
build

**Comment by wangjoshuah** [comment]
build

File: core-api/src/test/java/com/optimizely/ab/OptimizelyTest.java [hunk 4]
```diff
@@ -3495,6 +3496,36 @@ public void isFeatureEnabledWithExperimentKeyForcedOfFeatureEnabledFalse() throw
         assertFalse(optimizely.isFeatureEnabled(FEATURE_FLAG_MULTI_VARIATE_FEATURE.getKey(), testUserId));
     }
 
+    /**
+     * Verify that the {@link Optimizely#activate(String, String, Map<String, String>)} call
+     * uses forced variation to force the user into the second variation in which FeatureEnabled is not set
+     * feature enabled will return false by default
+     */
+    @Test
+    public void isFeatureEnabledWithExperimentKeyForcedWithNoFeatureEnabledSet() throws Exception {
```
Comment (mnoman09) [inline core-api/src/test/java/com/optimizely/ab/OptimizelyTest.java]: @thomaszurkan-optimizely  here is the test case I am checking in which we are not setting featureEnabled variable in datafile and in config so its returning false by default

**Comment by mikeng13** [comment]
build

**Pull request merged**